### PR TITLE
Allow external http client and add custom request options

### PR DESCRIPTION
**Pull request by Gitii**
Hi,

Thanks for creating this npm module and sharing it with the world!
I've used it and made a few small improvements:

* Aligned spacing & formatting (more by accident)
* Exported `HttpClient` and added a `clientFactory` option for `Kindle` class. That is very useful when you want to NOT use the `tls-client-api` server but the shared library.
* Added pagination handling to `Kindle` class. The default behaviour hasn't changed, there is an on-demand option for it.
* Moved fetching of books into a separate function (`fetchBooks`) to reduce complexity (increased due to pagination handling). 
* Added query and filtering options (also includes the pagination).

Please take a look at it and I hope that we merge them!

Thanks!

### CODE REVIEW

**Comment by Xetera** [comment]
This is excellent, thanks a lot for the PR. I've been meaning to make `tls-client-api` optional/more modular but just haven't had the motivation for it since I solved my own problem with the proxy.

One thought I have is since this includes pagination, I feel like it's the kind of thing that should be turned on by default. It feels a little bit strange to pass a false boolean flag to turn something on. Can we flip this behavior around to have pagination on by default so it can be released as the next major version to prevent breaking changes for existing users?

File: src/query-filter.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  /**
+   * Set the number of results to return.
+   * Default is 50.
+   */
+  querySize?: number;
+  /**
+   * When false, pagination is enabled and all results are returned.
+   * The default is false and only the first page of results is returned.
+   */
+  firstPageOnly?: boolean;
```
Comment (Xetera) [inline src/query-filter.ts]: I think this should read `The default is true`?

**Review by Xetera — CHANGES_REQUESTED** [review]
Looks good overall, if we can just get the default changed and add some documentation in the readme about pagination that would be enough

File: src/query-filter.ts (hunk 1 above)
Comment (Gitii) [reply src/query-filter.ts]: Ah, sorry. I've changed the field name and description several times. Nothing felt quite right... and apparently, I've missed this mistake:
Nevertheless, I've updated the PR with a refined version.

**Comment by Gitii** [comment]
@Xetera Do you have a preferred formatter / code style that I can use for this project? I am fighting with my vscode right now, because there are no project specific settings. And my defaults don't seem to match yours.

**Comment by Gitii** [comment]
And I also noticed that the package scripts are a bit messed up. `build:all` calls yarn but the package manager is `pnpm`.

**Comment by Xetera** [comment]
2 spaces, semi, trailing commas is good. It's basically the changes you've got in your PR so that's no problem. I'd also like to stick to pnpm. Sorry, just feeling the embarrassment of having what I thought would just be a personal project get used by others I guess 😮‍💨

**Comment by Gitii** [comment]
> 2 spaces, semi, trailing commas is good. It's basically the changes you've got in your PR so that's no problem. I'd also like to stick to pnpm. Sorry, just feeling the embarrassment of having what I thought would just be a personal project get used by others I guess 😮‍💨

No worries, this is our free time and not work 😁
`pnpm` is fine. 
I've pushed a new commit that fixes it.

I also got sidetracked a little bit and wrote some tests. [I will push that commit](https://github.com/Gitii/kindle-api/commit/086cf50b344596fd615543455277d84c889bf435), too. Please take a look at it. I use `msw` to mock the network requests. That way the real Api is not part of the test and the tests can be run by anyone.
I can revert that commit if you do not like it. But I think it's worth it.

**Comment by Xetera** [comment]
Looks great, I've used msw a couple times before and it's awesome so I don't mind it. Happy to merge this if you feel it's good to go

**Comment by Gitii** [comment]
I think the PR is ready.

**Pull request merged**